engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -453,42 +453,60 @@ private void validateConstraintsForCurrentGroup(BaseBeanValidationContext<?> val
 		final BeanMetaData<U> beanMetaData = valueContext.getCurrentBeanMetaData();
 		final Map<Class<?>, Class<?>> validatedInterfaces = new HashMap<>();
 
-		// evaluating the constraints of a bean per class in hierarchy, this is necessary to detect potential default group re-definitions
-		for ( Class<? super U> clazz : beanMetaData.getClassHierarchy() ) {
-			BeanMetaData<? super U> hostingBeanMetaData = constraintMetaDataManager.getBeanMetaData( clazz );
-			boolean defaultGroupSequenceIsRedefined = hostingBeanMetaData.isDefaultGroupSequenceRedefined();
-
-			// if the current class redefined the default group sequence, this sequence has to be applied to all the class hierarchy.
-			if ( defaultGroupSequenceIsRedefined ) {
-				Iterator<Sequence> defaultGroupSequence = hostingBeanMetaData.getDefaultValidationSequence( valueContext.getCurrentBean() );
-				Set<MetaConstraint<?>> metaConstraints = hostingBeanMetaData.getMetaConstraints();
-
-				while ( defaultGroupSequence.hasNext() ) {
-					for ( GroupWithInheritance groupOfGroups : defaultGroupSequence.next() ) {
-						boolean validationSuccessful = true;
-
-						for ( Group defaultSequenceMember : groupOfGroups ) {
-							validationSuccessful = validateConstraintsForSingleDefaultGroupElement( validationContext, valueContext, validatedInterfaces, clazz,
-									metaConstraints, defaultSequenceMember );
-						}
-						if ( !validationSuccessful ) {
-							break;
-						}
-					}
+		boolean defaultGroupSequenceIsRedefined = beanMetaData.isDefaultGroupSequenceRedefined();
+		if ( defaultGroupSequenceIsRedefined ) {
+			validateConstraintsForRedefinedGroupSequence( validationContext, valueContext, validatedInterfaces, beanMetaData.getBeanClass(), beanMetaData );
+			validationContext.markCurrentBeanAsProcessed( valueContext );
+		}
+		else {
+			validateConstraintsForSingleDefaultGroupElement( validationContext, valueContext, validatedInterfaces, beanMetaData.getBeanClass(), beanMetaData.getDirectMetaConstraints(), Group.DEFAULT_GROUP );
+			List<Class<? super U>> classHierarchy = beanMetaData.getClassHierarchy();
```

Comment:
I plan to make `BeanMetaData` an abstract class with regular and property holder impls. Then in case of property holder there will be nothing else but the property holder class in this list.

Also note the change itself. Before this we were accessing a map to get the metadata that we already have. With this we remove that one additional map access.

Reply:
Yeah ... Let me try to explain how I got here. The main "problem" that I was trying to "solve" was to reduce the number of places where `constraintMetaDataManager.getBeanMetaData( clazz );` is called. In case of property holders we cannot use such call And that's why for example in case of cascadebles I've moved such call into metadata class. In this particular place except the call for metadata itself we also access the metadata for the "most specific" class twice.

We already have the metadata from the context here - `BeanMetaData<U> beanMetaData = valueContext.getCurrentBeanMetaData()`. But then in the loop `for ( Class<? super U> clazz : beanMetaData.getClassHierarchy() )` we would get the beans class as a first element and try to get that same metadata again - `BeanMetaData<? super U> hostingBeanMetaData = constraintMetaDataManager.getBeanMetaData( clazz );` 

Hence to skip this additional metadata retrieval I've broken the method into two parts - first one will work with the existing "current" metadata `valueContext.getCurrentBeanMetaData()` and in the second part we iterate through the bean class hierarchy skipping the first element which is the bean class. 

I also remember running JMH tests for these changes and the results were almost identical.

In the end this change gives us:
- one additional call to metadata manager removed ( as metadata for that class is already cached it's basically removing one Map access call). Which doesn't give us much in terms of perfomance
- as we don't call the metadata manager to get metadata at this point, the code is working for the property holders. Because we build the correct metadata for them in corresponding value contexts instead of building it here in the `ValidtorImpl`

Now thinking about wrapping the `String mappingName` or beans class into wrapper ... we might end up not needing this at all. As we would then, most likely, have the `getClassHierarchy()` method return the wrappers. Or even maybe return the metadata right away, instead of simply returning classes.